Reintroduce CPU migration tests

valschneider opened the pull request.

derkling commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+
+        for task in self.rtapp_profile.values():
+            for phase_id, phase in enumerate(task.phases):
+                cpu_util[phase.cpus[0]][phase_id] += UTIL_SCALE * (phase.duty_cycle_pct / 100)
```
Should not be more safe as:
```python
+= (UTIL_SCALE * phase.duty_cycle_pct) / 100
```

valschneider replied:
Welcome to the wonderful world of Python3 where `20 / 100` == `0.2` (`20 // 100` == `0`)

derkling replied:
Yes, I guess it was P3 magic... however, we should maybe keep consistency with the Linux kernel world we come from, and the interpretation above is what we use in kernel space... just in case we want to be more friendly with some "low-level guys" looking at our tests ;)

valschneider replied:
Yeah I feel you, I hated the guts out of that division change when I started doing Python3 stuff. Thing is, you can't hide it all the times with parentheses, so might as well embrace it.

derkling commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+        df = self.trace.analysis.load_tracking.df_cpus_signals()
+        sw_df = self.trace.df_events("sched_switch")
+
+        phase_start = sw_df[sw_df.next_comm == list(self.rtapp_profile.keys())[0]].index[0]
```
`index.min()` ?

valschneider replied:
Same comment as above.

derkling commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+            start = phase_start + UTIL_AVG_CONVERGENCE_TIME_S
+            # Trim the end a bit, otherwise we could have one or two events
+            # from the next phase
+            end = phase_start + self.phases_durations[phase] * .9
```
Since we have:
```python
PHASE_DURATION_S = 3 * UTIL_AVG_CONVERGENCE_TIME_S
```
I would say it could be just good to set:
```python
end = phase_start + UTIL_AVG_CONVERGENCE_TIME_S
```

derkling replied:
Actually, giving the following code, where you keep computing (end-start), why not computing directly here:
```python
stable_interval = end - start
```
and use it in L826 ?

valschneider replied:
> Since we have:
> 
> ```python
> PHASE_DURATION_S = 3 * UTIL_AVG_CONVERGENCE_TIME_S
> ```
> 
> I would say it could be just good to set:
> 
> ```python
> end = phase_start + UTIL_AVG_CONVERGENCE_TIME_S
> ```

IMO the larger the interval is, the better it is in terms of averages.



> Actually, giving the following code, where you keep computing (end-start), why not computing directly here:
> 
> ```python
> stable_interval = end - start
> ```
> 
> and use it in L826 ?

Can do.

derkling commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+
+    @classmethod
+    def get_migration_cpus(cls, te):
+        cpu_capacities = te.plat_info["cpu-capacities"]
```
Wondering if we should add in some check/errors/failure in case this info is not available.
Not sure if we have any other pre-check before executing a test that all the required data are there... but I would expect in that case we expose somehow what are the requirements for a test... and we don't do that here AFAIKS.

douglas-raillard-arm replied:
That key should always be there since it is computed from the target. If it is not, it will fail, saying that cpu-capacities cannot be found anyway. If we can provide more information by failing early, we could add more pre-tests, but otherwise I don't really see the point (it will never be silent in any case).

valschneider replied:
Mmm yeah, I did add this hook: https://github.com/ARM-software/lisa/blob/next/lisa/tests/kernel/test_bundle.py#L264, but I haven't been a perfect citizen in that I don't think I've used it everywhere I should've...

derkling commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+    def get_migration_cpus(cls, te):
+        cpu_capacities = te.plat_info["cpu-capacities"]
+
+        return [cpu for cpu, capa in cpu_capacities.items()
```
why not just "cap" ? :)

valschneider replied:
Because otherwise my mind [gets sidetracked](https://www.youtube.com/watch?v=1dWIPUuTPB4) whenever I read this line.

derkling replied:
Fine... although I would have pick a different name (e.g. `cpu_cap`)...

derkling commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+        cpu_capacities = te.plat_info["cpu-capacities"]
+
+        return [cpu for cpu, capa in cpu_capacities.items()
+                if capa == max(cpu_capacities.values())][:2]
```
Can we the list comprehension and the return in two lines?

valschneider replied:
Isn't that already two lines? Or do you mean save the list in a variable and then do the return?

derkling replied:
the second... since with all those parenthesis my mind [gets sidetracked](https://www.youtube.com/watch?v=1dWIPUuTPB4) ...

derkling commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+
+        cpus = cls.get_migration_cpus(te)
+        if not len(cpus) == 2:
+            raise RuntimeError("This workload requires 2 CPUs of highest capacity")
```
Mmm... this will not work on tri-gear systems with just one big CPU.
In principle, we can run on 2 medium, provided we configure the tasks wrt their capacity... not sure, do we still have this change with the new wlgen class?

valschneider replied:
Yeah so that's directly copied from the test we have in master (not the sanity check, but the behaviour).

Offline discussion with @derkling suggests that we could just try to find 2 CPUs of same capacities and scale the rtapp workload using https://github.com/ARM-software/lisa/blob/next/lisa/tests/kernel/scheduler/eas_behaviour.py#L374

derkling commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+            # A task that will migrate to another CPU
+            profile["migrating"] += Periodic(
+                duty_cycle_pct=20, duration_s=cls.PHASE_DURATION_S,
+                period_ms=cls.TASK_PERIOD_MS, cpus=[cpus[i]])
```
can we pass in the calibration cpu here ?
In that case `get_migration_cpus()` could be update to return the highest capacity pair of CPUs in the system...

valschneider replied:
As discussed offline, I'd rather change the duty cycles than affect the calibration, because otherwise 20% duty_cycle != 20% utilization.

valschneider commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -203,7 +203,9 @@ def _test_task_signal(self, signal_name, allowed_error_pct,
-        signal_mean = signal_df[UTIL_AVG_CONVERGENCE_TIME_S:][signal_name].describe()['mean']
+        signal = signal_df[UTIL_AVG_CONVERGENCE_TIME_S:][signal_name]
+
+        signal_mean = area_under_curve(signal) / (signal.index[-1] - signal.index[0])
```
Indexes are monotonic (https://github.com/ARM-software/trappy/commit/7bb0bd9f29711e2a40e88fa33ae04fcb7df94e56), and I have no idea if Pandas/Numpy records this for optimizing min/max. So directly using an index seems like the better choice to me.

derkling replied:
Yes, we had a sanitization step somewhere in trappy which grants that... I just find the code above more "robust" and (maybe) a bit more easy to read... although I've also personally always used your same style. That was just to possibly consider to change our style... but not sure if the "robustness" argument is "robust" enough ;)

qais-yousef replied:
I prefer using pandas.Series.mean() function. I find the `area_under_curve()/duration` usage convoluted. Besides the behavior to re-sample and interpolate before it does the integration is semi magic. Most importantly around decay time where there's a big gap of no samples until the task wakes up again.

If it does fix a problem then we need to understand exactly why. Can you add a few debug messages to help with that? I suggest calculate using the 2 methods and print them out in a log.debug() or something.

By the way, why you had `___.describe()['mean']` instead of `___.mean()` in the removed line? Curious what's the difference.

valschneider replied:
> I prefer using pandas.Series.mean() function. I find the `area_under_curve()/duration` usage convoluted. Besides the behavior to re-sample and interpolate before it does the integration is semi magic. Most importantly around decay time where there's a big gap of no samples until the task wakes up again.
> 

That was my feeling as well, so I went for `mean()` until I hit issues and reconsidered (see below).

> If it does fix a problem then we need to understand exactly why. Can you add a few debug messages to help with that? I suggest calculate using the 2 methods and print them out in a log.debug() or something.
> 

I did that and the difference is noticeable. With the integral mean I can safely set a 5% threshold (usual error % is < 2%), with the regular mean I need at least 10%. I had a chat with @derkling and the integration *is* the way to go - IIRC that can be verified on the PELT simulator. He'll 100% explain it better than I, but that integration is required.

> By the way, why you had `___.describe()['mean']` instead of `___.mean()` in the removed line? Curious what's the difference.

There's none, that wasn't necessary :)

valschneider commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+
+    PHASE_DURATION_S = 3 * UTIL_AVG_CONVERGENCE_TIME_S
+    """
+    The duration of a single phase
```
Nope, docstrings are always below definitions.

douglas-raillard-arm replied:
autodoc gives the choice between comments above or a string literal below. We went the string way, to be consistent with functions/classes/modules
http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoattribute

valschneider commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+    **Design notes:**
+
+    Since we sum up the utilization of each task, make sure not to overload the
+    CPU - IOW, there should always be some idle cycles.
```
Yeah, maybe...

valschneider commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -735,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+
+        :returns: A dict of the shape {cpu : {phase_id : trace_util}}
+        """
+        cpu_util = {cpu : {phase_id : 0 for phase_id in range(self.nr_phases)}
```
I usually find pre-initing easier on the eyes than using `setdefault()`, but that's personal preference :)

valschneider commented:
So, I introduced some wilder changes:
* Added a `capacity-classes` entry in plat_info (@douglas-raillard-arm, is there a way to cleanly mark this as "derived" information? There's no need to serialize it as it's computed from CPU capacities)
* Moved `EASBehaviour.unscaled_utilization` to `RTATestBundle` to use it in `xCPUMigration`
* `xCPUMigration` now scans through `capacity-classes` to get 2 CPUs of the same capacity, and uses `unscaled_utilization`

douglas-raillard-arm commented:
> is there a way to cleanly mark this as "derived" information? There's no need to serialize it as it's computed from CPU capacities

If that is always derived from there, that should probably turned into a function that derives the information from the key that already exists. Having keys derived from other keys brings coherence issues. The user has final control of every key, so it's better to keep the set of controls as small as possible, without redundance when possible (in that case, it is possible, unlike deriving everything from an EM that may not be present for example).

douglas-raillard-arm commented:
If we really want to have that computed information travel along the `MultiSrcConf` instances, we could introduce a way to define derived keys, that won't be settable from the file and that will not be dumped either. This way, we avoid confusing state duplication, and we can fully reflect that in the documentation. Its source name could be derived from the source names of all "mother" keys, so we still keep track of where things are coming from.

derkling commented:
I like better the first solution proposed by @douglas-raillard-arm ... if something is derived, what can be a good reason to not compute it using just a (memoized) function ?

valschneider commented on lisa/tests/kernel/scheduler/eas_behaviour.py:
```diff
@@ -431,7 +393,7 @@ def test_task_placement(self, energy_est_threshold_pct=20, nrg_model:EnergyModel
     def get_rtapp_profile(cls, te):
         # 50% of the smallest CPU's capacity
-        duty = cls.unscaled_utilization(cls.min_cpu_capacity(te), 50)
+        duty = cls.unscaled_utilization(te, te.plat_info["capacity-classes"][0][0], 50)
```
Yeeeeah when I wrote that change it striked me as extremely newcomer-unfriendly.

You have to read that `plat_info` access as "The first CPU of the first CPU capacity class" (they're sorted, so it's the group CPUs with smallest capacity).

I could turn that into
```python
@classmethod
def smallest_cpus(te):
    return te.plat_info["capacity-classes"][0]

duty = cls.unscaled_utilization(te, cls.smallest_cpus(te)[0], 50)
```

And write its sibling for biggest CPUs.

valschneider commented on lisa/tests/kernel/scheduler/load_tracking.py:
```diff
@@ -733,3 +735,244 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+                trace_metrics[cpu_str].data[phase_str] = TestMetric(
+                    trace_cpu_util[cpu][phase])
+                deltas[cpu_str].data[phase_str] = TestMetric(
+                    -100 * (expected_cpu_util[cpu][phase] - trace_cpu_util[cpu][phase]) /
```
You'll love the next version ;)

valschneider commented:
Updated using derived keys.

douglas-raillard-arm commented on lisa/platforms/platinfo.py:
```diff
@@ -16,15 +16,37 @@
+    capacities = sorted(set(cpu_caps.values()))
+
+    classes = [
+        [cpu for cpu, cpu_capa in cpu_caps.items() if cpu_capa == capacity]
```
The inner list comprehension needs to be sorted as well, as it depends on the iteration order of a dictionary, which cannot be relied upon.

Also, this can be simplified using:
```
def compute_capa_classes(conf): 
     cpu_caps = conf['cpu-capacities']
     return [
         sorted(cpu_group)
         for capa, cpu_group in utils.groupby(cpu_caps.items(), key=operator.itemgetter(1))
      ]
```
`operator.itemgetter()` just returns a function that will get the `N`th item of a sequence, in this case the 2nd item of the 2-tuples returned by `dict.items()`
Could be even worth adding a `group_by_value()` helper in `utils`:
```
def group_by_value(mapping, key_sort=lambda k: 0):
     return {
        val: sorted((k for k, v in key_group), key=key_sort)
         for val, key_group in groupby(mapping.items(), key=operator.itemgetter(1))
     }
```
And then:
```
def compute_capa_classes(conf): 
     return classes = [*group_by_value(conf['cpu-capacities'], key_sort=lambda k: k).values()]
```

douglas-raillard-arm commented on lisa/tests/kernel/scheduler/misfit.py:
```diff
@@ -127,16 +127,7 @@ def __init__(self, res_dir, plat_info, rtapp_profile):
-
-        capacities = sorted(cpu_classes.keys())
-        self.cpu_classes = [cpu_classes[capacity] for capacity in capacities]
+        self.cpu_classes = plat_info['capacity-classes']
```
may be better to refer to self.plat_info where it is actually needed. That allows easier refactoring (the code just depends on a `plat_info` to work, instead of a specific class), and allows better tracking what data is used as the input of each test by looking at the debug log (which logs access to `plat_info`).

valschneider replied:
I feel much more comfortable storing this value into a variable, especially since it's only ever used in this method - I now see it's only used in the `__init__()` so I could make it a local variable instead of an instance attribute .

douglas-raillard-arm replied:
right, if it's just some local use then there is no issue. What we just want to avoid is ending with quite generic code, that ends up seemingly deeply tied to a class where in fact, it could easily be extracted and put in a helper (or a base class) if needed.

douglas-raillard-arm commented on lisa/tests/kernel/scheduler/misfit.py:
```diff
@@ -127,16 +127,7 @@ def __init__(self, res_dir, plat_info, rtapp_profile):
         super().__init__(res_dir, plat_info, rtapp_profile)
 
         cpu_capacities = plat_info['cpu-capacities']
```
this line can be removed as well

valschneider commented:
@douglas-raillard-arm All done

valschneider commented:
@derkling, wanna re-re-review this one? :D
